documentation/api/authentication.md

```diff
@@ -0,0 +1,163 @@
+
+### Getting the Login.gov Token
+Using the code we received back from login.gov, we call their [token endpoint](https://developers.login.gov/oidc/token/).
+To call the endpoint we also need a [client_assertion jwt](https://developers.login.gov/oidc/token/#client_assertion)
+which is created from a private key we generated, and gave Login.gov the public key certificate.
+
+If there is any error with [the response](https://developers.login.gov/oidc/token/#token-response) from login.gov, error.
+
+We parse the JWT, validating it against the [public keys](https://developers.login.gov/oidc/certificates/)
+provided by login.gov. We refresh these in the event we don't find a matching key based on the KID (key ID).
+
+We additionally validate that the token is valid across several other fields (expired, audience, issuer, etc), but
+most importantly we check that the `nonce` value in the key matches the one we originally sent to login.gov.
+If any of these validations fail, we error.
+
+### Creating our own JWT
+
+If all validation passed, we'll then check if the `sub` value from the token matches
+any user in `link_external_user` table. If not, we'll create a record + create a user.
+In either case, we'll update the email from login.gov in this table.
+
+We create a `user_token_session` object in our database which has a `token_id`, `expires_at` timestamp, and `is_valid` boolean.
+
+The expiration of the token we create is not managed in the token itself, but instead in our database, and can be managed
+by our logout and refresh endpoints.
+
+With this session token we generate a JWT containing the `token_id`, `user_id` and a few miscellaneous JWT attributes.
+
+### Redirect to final destination
+With the token we redirect to the configured final destination URL with a message, token, and whether the user was new.
+
+In future iterations we may adjust this logic to allow for the final destination to be set by the caller in the /users/login endpoint, however
+for security reasons we'll want to limit where we can send to.
+
+## Error handling
+Because these aren't standard GET/POST/etc. endpoints and instead handle
+all responses with redirects, we redirect to the same final destination in the event
+an error occurs. We send an `error` and `error_description` parameter with rough
+information about what issue occurred.
+
+In both our login and callback endpoints we wrap the entire endpoint in a try/catch
+block that handles catching errors and doing this redirect to the final destination.
+
+Most errors fall into two categories:
+* Misconfiguration in the API
+* Someone is calling the callback URL directly with random data
+
+Because we need to protect ourselves against both of these, many error scenarios
+are actually not something we would ever expect a user to hit.
+
+## Logout
+The logout endpoint will log out the session associated with the passed-in token.
+
+It does this by setting the `is_valid` boolean to False.
```

Comment:
not directly related, but curious if there's been a decision about how often to clean out session related tables

Reply:
Haven't had a discussion, I'd argue that it would be good to keep sessions for at least a few weeks for debugging purposes. Unless we're getting into the tens of millions, I wouldn't see much concern for performance.

Maybe something like 6 months? When we launch this, we'll get a better idea of how many sessions we're maintaining and go from there probably